src/metric/metric.cpp

```diff
@@ -14,53 +14,130 @@
+      return new MAPEMetric(config);
+    } else if (type == std::string("gamma")) {
+      Log::Warning("Metric gamma is not implemented in cuda_exp version. Fall back to evaluation on CPU.");
+      return new GammaMetric(config);
+    } else if (type == std::string("gamma_deviance")) {
+      Log::Warning("Metric gamma_deviance is not implemented in cuda_exp version. Fall back to evaluation on CPU.");
+      return new GammaDevianceMetric(config);
+    } else if (type == std::string("tweedie")) {
+      Log::Warning("Metric tweedie is not implemented in cuda_exp version. Fall back to evaluation on CPU.");
+      return new TweedieMetric(config);
+    }
+  } else {
+  #endif  // USE_CUDA_EXP
+    if (type == std::string("l2")) {
+      return new L2Metric(config);
+    } else if (type == std::string("rmse")) {
+      return new RMSEMetric(config);
+    } else if (type == std::string("l1")) {
+      return new L1Metric(config);
+    } else if (type == std::string("quantile")) {
+      return new QuantileMetric(config);
+    } else if (type == std::string("huber")) {
+      return new HuberLossMetric(config);
+    } else if (type == std::string("fair")) {
+      return new FairLossMetric(config);
+    } else if (type == std::string("poisson")) {
+      return new PoissonMetric(config);
+    } else if (type == std::string("binary_logloss")) {
+      return new BinaryLoglossMetric(config);
+    } else if (type == std::string("binary_error")) {
+      return new BinaryErrorMetric(config);
+    } else if (type == std::string("auc")) {
+      return new AUCMetric(config);
+    } else if (type == std::string("average_precision")) {
+      return new AveragePrecisionMetric(config);
+    } else if (type == std::string("auc_mu")) {
+      return new AucMuMetric(config);
+    } else if (type == std::string("ndcg")) {
+      return new NDCGMetric(config);
+    } else if (type == std::string("map")) {
+      return new MapMetric(config);
+    } else if (type == std::string("multi_logloss")) {
+      return new MultiSoftmaxLoglossMetric(config);
+    } else if (type == std::string("multi_error")) {
+      return new MultiErrorMetric(config);
+    } else if (type == std::string("cross_entropy")) {
+      return new CrossEntropyMetric(config);
+    } else if (type == std::string("cross_entropy_lambda")) {
+      return new CrossEntropyLambdaMetric(config);
+    } else if (type == std::string("kullback_leibler")) {
+      return new KullbackLeiblerDivergence(config);
+    } else if (type == std::string("mape")) {
+      return new MAPEMetric(config);
+    } else if (type == std::string("gamma")) {
+      return new GammaMetric(config);
+    } else if (type == std::string("gamma_deviance")) {
+      return new GammaDevianceMetric(config);
+    } else if (type == std::string("tweedie")) {
+      return new TweedieMetric(config);
+    }
+  #ifdef USE_CUDA_EXP
   }
+  #endif  // USE_CUDA_EXP
```

Comment:
can we reduce these duplicated codes? like use predefined macros.

Reply:
In the next stage, we need to merge the CUDA version of the metrics one by one. I think writing explicitly for each metric and for both `cuda_exp` and other versions is more convenient for the next stage.